### PR TITLE
Add cosine-distance meta example

### PR DESCRIPTION
**Pull request by IOcodegeass**
Added **cosine.sg** and **cosine.rst** to create meta examples for distance-cosine and generated cookbook for distance-cosine.

### CODE REVIEW

**Comment by karlnapf** [comment]
@vigsterkr The cookbook preview does not show the right commit version for some reason. I noticed this before: the added cookbook is not part of the html renderer

**Comment by karlnapf** [comment]
@IOcodegeass can you squash the commits please

File: doc/cookbook/source/examples/distance/cosine.rst [hunk 1]
```diff
@@ -0,0 +1,40 @@
+==================
+Cosine	 Distance
+==================
+
+The Cosine distance for real valued features is the similarity between two vectors by finding their angle.
+
+.. math::
+
+    d({\bf x},{\bf x'})= 1-\frac{\sum_{i=1}^{n}{\bf x_i}.{\bf x'_i}}{\sqrt{\sum_{i=1}^{n}{\bf {x_i}^2}\sum_{i=1}^{n}{\bf {x'_i}^2}}}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/distance/cosine.rst]: Can you use $x^\top x$ for vector dot products?

Comment (karlnapf) [inline doc/cookbook/source/examples/distance/cosine.rst]: and $d$ for the dimensions

File: doc/cookbook/source/examples/distance/cosine.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+Imagine we have files with data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) as
+
+.. sgexample:: cosine.sg:create_features
+
+We create an instance of :sgclass:`CCosineDistance` by passing it :sgclass:`CDenseFeatures`.
+
+.. sgexample:: cosine.sg:create_instance
+
+Distance matrix can be extracted as follows:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/distance/cosine.rst]: The distance matrix

File: doc/cookbook/source/examples/distance/cosine.rst [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+.. sgexample:: cosine.sg:extract_distance
+
+We can use the same instance with new :sgclass:`CDenseFeatures` to compute distance.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/distance/cosine.rst]: We can compute non-symmetric distances via

**Comment by karlnapf** [comment]
About the failing octave build, you read the error message. It might have told you that the build timed out, I restarted it :)

File: doc/cookbook/source/examples/distance/cosine.rst (hunk 1 above)
Comment (IOcodegeass) [inline doc/cookbook/source/examples/distance/cosine.rst]: u mean d instead on n for dimension of vector. But that would match with 'd' for distance measure.

Comment (IOcodegeass) [inline doc/cookbook/source/examples/distance/cosine.rst]: Is this correct ?

d({\bf x},{\bf x'})= 1-\frac{{\bf x^\top x'}}{\sqrt{\sum_{i=1}^{d}{\bf {x_i}^2}\sum_{i=1}^{d}{\bf {x'_i}^2}}}

File: doc/cookbook/source/examples/distance/cosine.rst (hunk 2 above)
Comment (IOcodegeass) [inline doc/cookbook/source/examples/distance/cosine.rst]: Updated.

**Comment by vigsterkr** [comment]
@karlnapf infra contains the code

**Comment by karlnapf** [comment]
About the `d` overloading, just remove the d for the distance. It is not used anywhere anyways. Just give the formula on the right hand side of the = sign

**Comment by IOcodegeass** [comment]
Changes made in cosine.rst and cosine.sg.

File: doc/cookbook/source/examples/distance/cosine.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
+==================
+Cosine	 Distance
+==================
+
+The Cosine distance for real valued features is the similarity between two vectors by finding their angle.
+
+.. math::
+
+    1-\frac{{\bf x^\top x'}}{\sqrt{\sum_{i=1}^{d}{\bf {x_i}^2}\sum_{i=1}^{d}{\bf {x'_i}^2}}}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/distance/cosine.rst]: I just rendered the math here. It needs some minor tweaks (sorry for not checking earlier)
the _i subscript should not be in the bf macro, i.e. \bf {x}_i

Can you make write the denominator as `\Vert \bf{x}\Vert_2 \Vert \bf{x'}\Vert_2` and then say " where :math:`\Vert \cdot\Vert_2` is the Euclidean norm.

**Comment by karlnapf** [comment]
One more amend and ready to be merged

**Comment by IOcodegeass** [comment]
@karlnapf changes made. kindly review.

File: doc/cookbook/source/examples/distance/cosine.rst [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+Cosine	 Distance
+==================
+
+The Cosine distance for real valued features is the similarity between two vectors by finding their angle.
+
+.. math::
+
+    1-\frac{{\bf x^\top x'}}{\Vert \bf{x}\Vert_2 \Vert \bf{x'}\Vert_2 }
+
+where where :math:`\Vert \cdot\Vert_2` is the Euclidean norm, :math:`\bf x` and :math:`\bf x'` are :math:`d` dimensional feature vectors.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/distance/cosine.rst]: now you can actually remove the "d  dimensional" -- d is never used anywhere. Just say in the first sentence

"The Cosine distance for real valued features x and x' is the similarity as measured by their angle." And then remove the ":math:`\bf x` and :math:`\bf x'` are :math:`d` dimensional feature vectors."

**Comment by karlnapf** [comment]
Cool I merged the data, so you can update the data version as well, see wiki for testing data howto

**Comment by IOcodegeass** [comment]
@karlnapf updated **cosine.rst** .

**Comment by IOcodegeass** [comment]
@karlnapf I am not sure about [last commit](https://github.com/shogun-toolbox/shogun/pull/3567/commits/115616a76e6ba075326e64be0141a3ba61155f83). I have read the DEVELOPING.md and accordingly done         
`git commit data -m "Added testing data for cosine distance"
git push origin feature/cosine_distance`

But the explanation in DEVELOPING.md says to commit the updated version hash of the submodule which i couldn't understand.

**Comment by karlnapf** [comment]
It is exactly that. Now we wait for travis to execute the integration test for your example. Ping me if it passed.

**Comment by IOcodegeass** [comment]
@karlnapf CI is completed :)

**Comment by karlnapf** [comment]
Nice one. Well done on first merged cookbook page :)

**Comment by karlnapf** [comment]
Ah can you pls check to remove a potential cosine distance example from `undocumented/python_modular` folder

**Comment by IOcodegeass** [comment]
@karlnapf there exists euclidean distance example in ` undocumented/python_modular` folder although meta example for euclidean distance exists.

Should i remove that as well?

**Comment by karlnapf** [comment]
Yes! Thanks

**Pull request merged**